torch/_inductor/decomposition.py

```diff
@@ -680,3 +680,34 @@ def put(self, index, source, accumulate=False):
 def put_(self, index, source, accumulate=False):
     out = aten.put(self, index, source, accumulate=accumulate)
     return self.copy_(out)
+
+
+@register_decomposition(aten.index_reduce)
+def index_reduce(
+    self, dim: int, index, src, reduction_type: str, *, include_self: bool = True
+):
+    repeats = self.shape[dim + 1 :].numel() * self.shape[:dim].numel()
+    index_shape = (index.numel(), *self.shape[dim + 1 :], *self.shape[:dim])
+    perm = (*range(self.ndim - dim, self.ndim), 0, *range(1, self.ndim - dim))
+    scatter_index = index.repeat_interleave(repeats).reshape(index_shape).permute(perm)
+    return self.scatter_reduce(
```

Comment:
For certain inputs `scatter_reduce` falls back to eager, in which case this ends up creating a massive intermediate index tensor which isn't ideal. 
https://github.com/pytorch/pytorch/blob/c7bf5871ceb02321ec8d8fd5fd305ad9c2fd8132/torch/_inductor/lowering.py#L3420-L3431
Thankfully for inductor-specific decompositions you can conditionally not decompose by returning `NotImplemented`. So I would refactor `scatter_fallback` to extract the conditions under which it falls back and not decompose in those cases.

Reply:
Thanks for the hint! I have refactored parts of `scatter_fallback` into `use_scatter_fallback` that is used both in lowering and decomposition of relevant ops.